docs/design/Chem Integration.md

```diff
@@ -0,0 +1,66 @@
+```
+
+### Decoupling the data
+Initially, any data needed will be provided with the extension. This data would be what is used to map the fingerprints to actual packages. They often range around the 100mb-200mb range for useful data, but could be potentially larger if desired.
+
+However, it may be nice to decouple the data from the actual extension itself so they can be updated independently. In order to support his, Upgrade Assistant should provide a way for steps/extensions to provide step-dependent data. This could be something that a step could register it as having, and then Upgrade Assistant could provide commands such as the following:
+
+- `upgrade-assistnat manage-data list`: List registered data sources
+- `upgrade-assistant manage-data update --name [name]`: Update a registered data source (managed by the step itself)
+
+There are potentially many questions here as to handle the data. This is provided solely as a way to demonstrate a potential way to decouple the data required. Initially, the data will be included with the extension, but potentially have optional indexes (or indexes that carry more data and are much larger) will require some sort of way to manage the data.
+
+## Challenges
+
+There are a number of challenges with this, as it is completely based on heuristics in order to match package identities. Potential limitations:
+
+- If a package was never released to NuGet, it will never show up in the available indexes
+- Private builds of projects may not match external builds
+- Currently only planning on supporting packages from NuGet.org so loose assemblies from a company's internal teams will not be found.
+
+## Planned Work
+
+Below are the main work items that will be explored broken into two sprints. There is currently a pre-analysis migration effort being worked on, and this work may be updated to ensure that the information flows as expected into the reports generated there as those features become available.
+
+### Sprint 1
+- As a user, I can consume an extension that integrates chem (#482)
+- As a user, the chem integration can identify what loose assemblies I have (#483)
+- As a user, the chem integration will replace my loose assemblies with ones it identifies (#484)
+- As a user, I know which loose assemblies do not have any available matches (#485)
+
+### Sprint 2
+- As a user, I can update my chem data index independently of the extension (#486)
+- As a user, I can have access to the same data from multiple projects (#487)
+- As a user, I can receive recommendations from chem using multiple indexes (#488)
+
+## Out of scope/Future Work
+
+Some work that are currently not in scope:
+
+- Custom feeds (such as AzDO, Artifactory, file system, etc). NuGet.org will be the focus of the initial spike
+- Identifying any security concerns with switching to NuGet packages. We may want to explore identifying if a package has a security concern, but it's still probably the same binary.
```

Comment:
This feels low priority as it would be uncommon (though not impossible) for an old loose binary to be secure but the latest packaged version of a binary to be insecure.

Reply:
There are situations where the content of the library end up in a random nuget package that then was de-listed due to security concerns. But we will def need more data to know the % of such occurrences to make this feature a priority.

Reply:
I think a principle we're dancing around here is that we don't want to ever recommend moving to a package that has a known security vulnerability. The question is, what role should these indexes play in that? And I think this statement here is just saying that this is out of scope at this time from the perspective of the integrating the package indexes.

Reply:
That makes sense. The data collected by Chem doesn't have information on whether packages have security vulnerabilities I take it? In general, having a database like that somewhere for UA to use would be useful. I agree that it doesn't seem in scope for this work, though.

Reply:
Yup I agree with all that is said here. It makes sense potentially as as a separate step for Upgrade Assistant to analyze for any known CVE's, but didn't want to put it in scope here.